Add fields to group_trigger and group_member modules

zgalor opened the pull request:
Add the auto_resolve field in the module and example playbook.

@dkorn @cben PTAL

zgalor committed:
Add auto_resolve field to group_trigger module to enable auto_resolve

cben commented:
Naive question: how does this relate to AUTORESOLVE trigger mode on conditions?

zgalor commented:
you can set it on the condition, but if you don't enable it on the trigger it wouldn't work.
IIUC the auto_resolve field of the trigger determines the alert state:
will it fire on every new metric that satisfies the condition, or will it fire once, and then enter auto-resolving state until a resolving metric arrives.

dkorn commented on library/hawkular_alerts_group_trigger.py:
```diff
@@ -42,6 +42,11 @@
     required: False
+  auto_resolve:
+    description:
+      - is auto-resolve enabled
```
I'd add/change: "`, meaning switch to auto-resolve mode after firing"`

dkorn reviewed (commented):
@zgalor Thanks for the contribution!

Please consider adding also:
* `auto_resolve_alerts` - Resolve all unresolved alerts when auto-resolve condition-set is satisfied
* `auto_resolve_match` - The policy used for deciding whether the trigger auto-resolved condition-set is satisfied. ALL conditions must evaluate to true or ANY one condition must evaluate to true.

Both are needed and can be part of this PR.

zgalor commented:
@dkorn I suggest we add those in a different PR, just for the sake of time. 
I will not get to it before Wednesday, and we should at least have the auto-resolve feature merged.

zgalor committed:
Add fields to group member and group trigger

In group trigger added: event_text, tags
In group member added: description, tags

zgalor commented:
@cben @dkorn new fields added to allow collection in manageIQ - please review

dkorn commented on library/hawkular_alerts_group_member.py:
```diff
@@ -40,11 +40,20 @@
+  tags:
+    description:
+      - Trigger tags for member Trigger
+    required: False
```
Can you specify that a tag is a [name, value] pair? (like in the trigger module)

Please add `default: null`

dkorn commented on library/hawkular_alerts_group_member.py:
```diff
@@ -210,7 +224,9 @@ def main():
             name=dict(required=False, type='str'),
+            description=dict(required=False, type='str'),
             data_id_map=dict(required=False, type='dict'),
+            tags=dict(required=False, type='dict'),
```
Looking at the Hawkular Alerts docs [1], it looks like `memberDescription` and `memberTags` are required (as well as `memberContext`).
It might be a recent change, otherwise, I'm not sure how this worked.

[1] http://www.hawkular.org/docs/rest/rest-alerts.html#GroupMemberInfo

zgalor replied:
yes, i spoke with @lucasponce. He said it's only a documentation error, and he will fix it.

dkorn commented on library/hawkular_alerts_group_member.py:
```diff
@@ -222,18 +238,20 @@ def main():
+    group_id           = module.params['group_id']
+    member_id          = module.params['id']
+    member_name        = module.params['name']
+    member_description = module.params['description']
```
I would drop the `member` prefix from both `member_description` and `member_tags`
(I know it `member_name` has this prefix as well, also unneeded)

dkorn commented on library/hawkular_alerts_group_trigger.py:
```diff
@@ -355,7 +392,7 @@ def main():
 
     if state == "present":
-        res_args = hawkular_alerts.create_or_update_group_trigger(name, group_id, severity, enabled, conditions)
+        res_args = hawkular_alerts.create_or_update_group_trigger(name, event_text, group_id, severity, auto_resolve, tags, enabled, conditions)
```
please move `event_text` after `group_id`, as the latter is required.

dkorn commented on library/hawkular_alerts_group_trigger.py:
```diff
@@ -184,23 +204,29 @@ def conditions_update_required(self, trigger_id, desired_conditions):
         # `is not None` check verifies that omitted module params will not be updated
         if (name is not None and trigger.name != name):
             updates["name"] = name
+        if (event_text is not None and trigger.event_text != event_text):
```
As this grows (and will continue to) I think we should pass a dictionary of fields and iterate over them. something like:
```python
for attribute in group_trigger_attributes:
    if (attribute is not None and getattr(trigger, attribute) != attribute):
        updates[attribute] = attribute

```
@cben what do you think?

cben replied:
Agree :+1:.

dkorn reviewed (changes requested):
Few minor comments and suggestions

zgalor commented:
@dkorn I made the changes you requested and retested. Please review

cben commented on library/hawkular_alerts_group_trigger.py:
```diff
@@ -184,23 +204,29 @@ def conditions_update_required(self, trigger_id, desired_conditions):
 
             Returns:
-                Empty Hash (None) - If the name, severity and enabled status passed
+                Empty Hash (None) - If the name, event_text, severity, auto_resolve, tags and enabled status passed
```
what does "Empty Hash (None)" mean?  `{} != None`.  Just say "Empty Hash".

I'd also stop listing here (twice) all the attributes compared, just say that all values match.

zgalor commented:
@dkorn @cben I applied your recommendation and generalized `required_updates` to loop through dictionary instead of specific `if`s.

Tested locally

dkorn commented on library/hawkular_alerts_group_trigger.py:
```diff
@@ -284,10 +303,16 @@ def create_or_update_group_trigger(self, name, group_id, severity, enabled, cond
+                                             "severity": severity,
+                                             "auto_resolve": auto_resolve,
+                                             "tags": tags,
+                                             "enabled": enabled})
```
I'd probaby change the dict creation to:

```python
dict([(i, locals()[i]) for i in ('name', group_id', auto_resolve', 'event_text', 'severity', 'enabled', 'tags')])
```
or
```python
attrs_to_compare = dict()
for i in ('name', group_id', auto_resolve', 'event_text', 'severity', 'enabled', 'tags'):
    attrs_to_compare[i] = locals()[i]
```

not a blocker
